### PR TITLE
Version 3 – the collected changes in one pr

### PR DESCRIPTION
**Pull request by sprain**
This pull requests combines a number of changes which will result in becoming v3 of this library.

While there are BC breaks, most users will need to do no changes in their code if the library was already used as intended.

* PHP 7.4 is now required. (#101)
* All constructors are now `private`. Factory methods must be used. (#101)
* All classes are now `final`. (#119)
* Classes which are not supposed to be used outside this library are now marked with `@internal`. (#120)
* Removed deprecated `QrBill::setErrorCorrectionLevel()` method. (#102)

### CODE REVIEW

**Review by kohlerdominik — APPROVED** [review]
Our Integration Tests succeeded, so i think compatibility is fine. Comments of minor issues added in the code.

File: src/QrBill.php [hunk 1]
```diff
@@ -21,45 +22,35 @@
...
     /** @var AlternativeScheme[] */
-    private $alternativeSchemes = [];
+    private array $alternativeSchemes;
 
-    /** @var string  */
-    private $errorCorrectionLevel = self::ERROR_CORRECTION_LEVEL_MEDIUM;
+    private function __construct(Header $header)
+    {
+        $this->header = $header;
+        $this->creditorInformation = null;
```
[the remark this answers is not in the document]
Comment (sprain) [reply src/QrBill.php:46]: I'll have to check again whether there was a reason for that ;)

**Comment by sprain** [comment]
> Our Integration Tests succeeded, so i think compatibility is fine. Comments of minor issues added in the code.

@kohlerdominik Great feedback, thanks a lot! I will take care of the fixes-

**Comment by kohlerdominik** [comment]
@sprain are you going to merge this, so we can rebase #106 and make a PHP8-compatible release?

**Comment by sprain** [comment]
@kohlerdominik I am back from holidays and will merge and release this within the next few days.

**Comment by sprain** [comment]
🥳 https://github.com/sprain/php-swiss-qr-bill/tree/v3.0

**Pull request merged**